src/org/rascalmpl/repl/TerminalProgressBarMonitor.java

```diff
@@ -117,87 +118,80 @@ public void write(String str, int off, int len) throws IOException {
-         * all we have to do now is figure out where the last newline character is.
+         * new (unfinished) sentences. 
          * 
+         * The resulting buffer nevers contain any newline character.
          */
-        private void add(byte[] newInput, int offset, int len) {
+        private void store(byte[] newInput, int offset, int len) {
             // first ensure capacity of the array
             if (curEnd + len >= curCapacity) {
-                var oldCapacity = curCapacity;
-                curCapacity *= 2; // this should not happen to often. we're talking a few lines of text.
-                byte[] tmp = new byte[curCapacity];
-                System.arraycopy(buffer, 0, tmp, 0, oldCapacity);
-                buffer = tmp;
+                curCapacity *= 2; 
+                buffer = Arrays.copyOf(buffer, curCapacity);
             }
 
             System.arraycopy(newInput, offset, buffer, curEnd, len);
             curEnd += len;
-            lastNewLine = startOfLastLine();
         }
 
         public void write(byte[] n, OutputStream out) throws IOException {
-            add(n, 0, n.length);
-            flushToLastLine(out);
+            write(n, 0, n.length, out);
         }
 
+        /**
+         * Main workhorse looks for newline characters in the new input.
+         *  - if there are newlines, than whatever is in the buffer can be flushed.
+         *  - all the characters up to the last new new line are flushed immediately.
+         *  - all the new characters after the last newline are buffered.
+         */
         public void write(byte[] n, int offset, int len, OutputStream out) throws IOException {
-            add(n, offset, len);
-            flushToLastLine(out);
-        }
-
-        private void flushToLastLine(OutputStream out) throws IOException {
-            if (lastNewLine != -1) {
-                // write everything out (except the last unfinished line), but including the last newline
-                out.write(buffer, 0, lastNewLine + 1);
-                
-                if (lastNewLine + 1 == curEnd) {
-                    // nothing left
-                    curEnd = 0;
-                    lastNewLine = -1;
-                }
-                else {
-                    // copy the last line that does not end with a newline
-                    curEnd -= lastNewLine;
-                    System.arraycopy(buffer, lastNewLine + 1, buffer, 0, curEnd);
-                    lastNewLine = -1; // no newline anymore
-                }
-            }
+            int lastNL = startOfLastLine(n, offset, len);
 
-            // otherwise we wait until the next input comes to be able to complete a line.
+            if (lastNL == -1) {
+                store(n, offset, len);
+            }
+            else {
+                flush(out);
+                out.write(n, offset, lastNL + 1);
+                store(n, offset, lastNL - offset);
```

Comment:
@jurgenvinju isn't this a bug? Shouldn't it be `lastNL` as argument instead of `offset?

Reply:
yes definitely a bug!